From 4810e18d1825032ef58c8f8f37f6698813963582 Mon Sep 17 00:00:00 2001
From: CI Assistant <yossi.ilkanaev@intel.com>
Date: Mon, 9 Jul 2018 07:42:55 +0200
Subject: [PATCH] Merge pull request #63 in SW_UGW/linux from
 bugfix/UGW_SW-26599-ugw8.1-er2-igmp-mld-snooping-is-not-working-in-multicast
 to master

* commit '79cde6963048a17e537ef67f088598db598bb951':
  UGW_SW-26599: Fix for Multicast Snooping issue on 8.x
---
 net/mcast_helper/mcast_helper.c | 19 +++++++++++--------
 1 file changed, 11 insertions(+), 8 deletions(-)

diff --git a/net/mcast_helper/mcast_helper.c b/net/mcast_helper/mcast_helper.c
index a3277ed1c604..25c61d6f7d5f 100644
--- a/net/mcast_helper/mcast_helper.c
+++ b/net/mcast_helper/mcast_helper.c
@@ -902,16 +902,17 @@ static void mcast_helper_delete_gitxmc_record(MCAST_MEMBER_t * gitxmc_rec,MCAST_
 {
 	unsigned int flag = 0;
 
-	if (gitxmc_rec == NULL)
-	  return;
-
-	if (gimc_rec->mc_stream.sIP.ipType == IPV4) {
-		if (mch_acl_enabled){
-			flag = mcast_helper_update_mac_list(gitxmc_rec,gimc_rec,macaddr,action);
-			mcast_helper_invoke_return_callback(gimc_rec->grpIdx,netdev,(MCAST_STREAM_t *)&(gimc_rec->mc_stream),flag, gitxmc_rec->macaddr_count);
-		}
+	if ((gitxmc_rec == NULL) || (gimc_rec == NULL))
+		return;
 
+#ifdef CONFIG_MCAST_HELPER_ACL
+	if ((mch_acl_enabled == 1) && (gitxmc_rec->aclBlocked == 1)) {
+		return;
 	}
+#endif
+
+	flag = mcast_helper_update_mac_list(gitxmc_rec,gimc_rec,macaddr,action);
+	mcast_helper_invoke_return_callback(gimc_rec->grpIdx, netdev, (MCAST_STREAM_t *)&(gimc_rec->mc_stream), flag, gitxmc_rec->macaddr_count);
 }
 
 /*=============================================================================
@@ -2113,6 +2114,8 @@ int mcast_helper_sig_check_update_ip(struct sk_buff *skb)
 #ifdef CONFIG_MCAST_HELPER_ACL
 					gitxmc_rec->aclBlocked = 0;
 #endif
+					flag = mcast_helper_update_mac_list(gitxmc_rec,gimc_rec,NULL,MC_F_ADD);
+					mcast_helper_invoke_return_callback(gimc_rec->grpIdx,gitxmc_rec->memDev,(MCAST_STREAM_t *)&(gimc_rec->mc_stream),flag, gitxmc_rec->macaddr_count);
 
 				}
 
